File: server/org.eclipse.glsp.example.workflow/src/main/java/org/eclipse/glsp/example/workflow/provider/AbstractNextOrPreviousNavigationTargetProvider.java
/********************************************************************************
 * Copyright (c) 2020 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
package org.eclipse.glsp.example.workflow.provider;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import org.eclipse.glsp.api.model.GraphicalModelState;
import org.eclipse.glsp.api.provider.NavigationTargetProvider;
import org.eclipse.glsp.api.types.EditorContext;
import org.eclipse.glsp.api.types.NavigationTarget;
import org.eclipse.glsp.api.utils.ClientOptions;
import org.eclipse.glsp.example.workflow.wfgraph.TaskNode;
import org.eclipse.glsp.graph.GEdge;

public abstract class AbstractNextOrPreviousNavigationTargetProvider implements NavigationTargetProvider {

   @Override
   public List<? extends NavigationTarget> getTargets(final EditorContext editorContext,
      final GraphicalModelState modelState) {
      Optional<String> sourceUri = ClientOptions.getValue(modelState.getClientOptions(), ClientOptions.SOURCE_URI);
      return editorContext.getSelectedElementIds().stream()
         .flatMap(id -> modelState.getIndex().get(id).stream())
         .filter(TaskNode.class::isInstance).map(TaskNode.class::cast)
         .flatMap(taskNode -> getEdges(taskNode, modelState).stream())
         .map(edge -> getSourceOrTarget(edge))
         .map(id -> new NavigationTarget(sourceUri.orElse(""), createElementIdMap(id)))
         .collect(Collectors.toList());
   }

   private Map<String, String> createElementIdMap(final String id) {
      Map<String, String> map = new HashMap<>();
      map.put(NavigationTarget.ELEMENT_IDS, id);
      return map;
   }

   protected abstract Collection<GEdge> getEdges(TaskNode taskNode, GraphicalModelState modelState);

   protected abstract String getSourceOrTarget(GEdge edge);
}

File: examples/workflow/workflow-glsp-server/src/provider/abstract-next-or-previous-navigation-target-provider.ts
/********************************************************************************
 * Copyright (c) 2022-2023 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
import { Args, EditorContext, NavigationTarget } from '@eclipse-glsp/protocol';
import { GEdge, ModelState, NavigationTargetProvider } from '@eclipse-glsp/server';
import { inject, injectable } from 'inversify';
import { TaskNode } from '../graph-extension';

@injectable()
export abstract class AbstractNextOrPreviousNavigationTargetProvider implements NavigationTargetProvider {
    abstract targetTypeId: string;

    @inject(ModelState)
    protected readonly modelState: ModelState;

    getTargets(editorContext: EditorContext): NavigationTarget[] {
        const sourceUri = this.modelState.sourceUri;
        const edges: GEdge[] = [];
        editorContext.selectedElementIds
            .map(id => this.modelState.index.get(id))
            .filter(e => e instanceof TaskNode)
            .map(e => e as TaskNode)
            .map(taskNode => edges.push(...this.getEdges(taskNode)));
        return edges.map(edge => this.getSourceOrTarget(edge)).map(id => this.createNavigationTarget(sourceUri, id));
    }

    protected createNavigationTarget(sourceURI: string | undefined, id: string): NavigationTarget {
        return { uri: sourceURI ?? '', args: this.createElementIdMap(id) };
    }

    protected createElementIdMap(id: string): Args {
        const map: Args = {};
        map[NavigationTarget.ELEMENT_IDS] = id;
        return map;
    }

    protected abstract getEdges(taskNode: TaskNode): GEdge[];

    protected abstract getSourceOrTarget(edge: GEdge): string;
}
